### ISSUE TITLE
Smeshing fails on restarting Node

### ISSUE THREAD

**Issue by brusherru** [spacemeshos/go-spacemesh#2858]
## Description
Smeshing fails after a successful setup on the next Node run.

After smeshing is set up the Node creates `postdata_0.bin` and `postdata_metadata.json` in the directory that the User defined as a `smeshing-datadir`.
On the next run Node tries to find `key.bin` in this directory instead of the previously used default location `~/post/data/key.bin`. As a consequence, it can't find it and creates a new `key.bin` there with the different (new) Node ID. Due to the new Node ID, it fails:
```
failed to complete post setup	{"node_id": "7ca3802629e39fe48e334e104b84590683db25dfe639afd92615719e6084dfd2", "module": "atxBuilder", "errmsg": "`ID` config mismatch; expected: 7ca3802629e39fe48e334e104b84590683db25dfe639afd92615719e6084dfd2, found: 8d46e54777dd4d60beeca9fd02b3ab1a5c2797af43d801844a3b22c71cc075a0, datadir: /some/user/defined/path/for/post/data", "name": "atxBuilder"}
```
Then if the User cleans up the defined post directory (or left only `key.bin` there) it will work fine.

## Steps to reproduce
1. Run the Node from scratch using some `node-config.json` without smeshing props
2. Run `smrepl --server localhost:9092`
3. Run `post setup`. Follow the steps, specify a non-default directory for the post data.
4. Put the new section `smeshing` in the `node-config`
4. Restart Node process

## Actual Behavior
Node isn't Smeshing. Post data "corrupted".
The Node log (related parts):
```
2021-10-04T20:27:01.205+0300	INFO	00000.defaultLogger	App version: v0.2.2-beta.1. Git: 38056f5-dirty - 38056f59d331e35e1e576c12040f522c610aff35 . Go Version: go1.15.13. OS: darwin-amd64 
...
2021-10-04T20:27:02.065+0300	INFO	starting spacemesh	{"data-dir": "/Users/brusher/Library/Application Support/Electron/node-data/205", "post-dir": "/Users/brusher/spacemesh", "hostname": "Kirill-557.local", "name": ""}
2021-10-04T20:27:02.065+0300	INFO	00000.defaultLogger	Looking for identity file at `/Users/brusher/spacemesh/key.bin`
2021-10-04T20:27:02.065+0300	INFO	00000.defaultLogger	Identity file not found. Creating new identity...
2021-10-04T20:27:02.066+0300	INFO	00000.defaultLogger	created new identity	{"public_key": "29137", "name": ""}
...
2021-10-04T20:27:02.399+0300	INFO	29137.clock        	started notifying	{"node_id": "29137508efea26a1777a84c7d13f53f7d287b8cb3cacf789c09c3a3d5dc2dd9e"}
2021-10-04T20:27:02.399+0300	INFO	29137.post         	post setup session starting	{"node_id": "29137508efea26a1777a84c7d13f53f7d287b8cb3cacf789c09c3a3d5dc2dd9e", "module": "post", "data_dir": "/Users/brusher/spacemesh", "num_units": "4", "labels_per_unit": "1024", "bits_per_label": "8", "provider": "1", "name": "post"}
...
2021-10-04T20:27:02.400+0300	INFO	00000.defaultLogger	starting new grpc server on :9092
2021-10-04T20:27:02.400+0300	ERROR	29137.atxBuilder   	failed to complete post setup	{"node_id": "29137508efea26a1777a84c7d13f53f7d287b8cb3cacf789c09c3a3d5dc2dd9e", "module": "atxBuilder", "errmsg": "`ID` config mismatch; expected: 29137508efea26a1777a84c7d13f53f7d287b8cb3cacf789c09c3a3d5dc2dd9e, found: 6859a5fe4f53226bd43c607053dcd62cb3bdf2eda5d9b01b4f1de6d1715d4e05, datadir: /Users/brusher/spacemesh", "name": "atxBuilder"}
...
```

## Expected Behavior
Smeshing works well on the second run without any tricks to make it work.

There we should think about how it should work:
- The NodeID / SmesherID should be the same and should be stored in some default location only (so we don't need to store it in post data dir and worry that someone can delete it).
   In case that we do not assume that NodeID and reward address should change depending on the open wallet — I suggest storing this file in a single and even more secure place than `~/post/data/key.bin`.
   E.G. `%APPDATA%/spacemesh/key.bin` (`~/Libraries/Application Support/spacemesh/key.bin` on macOS, `~/AppData/Roaming/spacemesh/key.bin`, `~/.config` on Linux)
   I think this is the best option, but I can miss something.
- The `key.bin` file should be copied to the post data-dir right on the setting up smeshing.

## Environment
macOS 10.13.6
[go-spacemesh v.0.2.2-beta1](https://github.com/spacemeshos/go-spacemesh/releases/tag/v0.2.2-beta.1)

**Comment by avive** [comment]
From product perspective I hope we can finally agree on the important requirement that the nodeId/smesherId of a node should never change unless user manually deletes the local node data and regardless of where the nodeId is persisted. For example, if user changes his post / smeshing settings in any way such as resizing post init date, stopping to smesh, changing rewards address, starting to again smesh, etc.. the node/smesher id that was generated on first node run should stay the same regardless of where node team decided to persist it. We should strive to simplify and reduce changing variable as much as possible to reduce perceptual complexity in a very complex platform. Having multiple smesher ids also makes it complicated to comprehend historical rewards, especially if user changed his rewards address. The nice-to-have feature that users should be able to get post data generated remotely should be designed in a way that doesn't break this basic requirement. For example, it can be implemented in a way that if user imports a post data folder that that includes a node key file than its node id changes for that id but only in that 'import' user case, but it shouldn't change otherwise.

**Comment by brusherru** [comment]
>  For example, it can be implemented in a way that if user imports a post data folder that that includes a node key file than its node id changes for that id but only in that 'import' user case, but it shouldn't change otherwise.

E.G. set post data-dir to `~/my/custom/post` and copy contents of the original `~/post` directory there and then it should use the same Smesher ID. Right?
In this case, we definitely should place `key.bin` in the post directory, but probably we should also store it somewhere in a "bullet-proof" place and fall back to it?

Btw, is there any reason to have two terms: "Node Id" and "Smesher Id"?
It seems that they are equal. Do we have some more difficult cases where they can differ?

**Comment by avive** [comment]
I had a chat with @noamnelke about this. There are multiple considerations here:
1. High-level and how things should work: 
- SmesherId/NodeId should really be called `PostId` and should be renamed to users as such in clients and dash/explore. The current name is misleading and was born out of historical misuse of this id (see bullet 2 below).
- It shouldn't exist before the user inits a post. 
- It should not change while the the node is smeshing across node's session (user didn't delete the post data).
- It should not change if user resizes an existing post.
- It should not change if user stopped smeshing and started smeshing again later using same post data.
- It is the context that post data, atxs and rewards are associated with.
2. Currently and for a long time this id has been misused by the node to be used as a unique node id in logs. There should be another random id generated on node first runtime session that should identify the node in the logs and in metrics independent of PostId.
3. The issue reported here is due to a bug related to the NodeId that will be patched soon so we can complete gpu-post basic flow for smapp 0.2. There's also a longer-term solution which involves the creation of an independent node id (bullet 2 above).

@noamnelke - please review this summary

**Comment by brusherru** [comment]
Right now calling `SmesherService/StopSmeshing` with `{ deleteFiles: true }` doesn't delete `key.bin`, only `post*` files.
So ID won't change even on deleting of post data (except manually cleaning post dir).
Is it correct?

**Comment by avive** [comment]
I think this is a bug. @noamnelke is aware of current issues and is aware that this stuff is blocking shipping 0.2 so I hope these will be fixed soon for most cases. Q: Is the happy flow of the basic use cases affected? e.g. Install smapp, run local node, start smeshing?

**Comment by avive** [comment]
@moshababo - this is issue is quite bad for users - it basically means smeshing will stop after you restart your node which is expected behavior. It will be great if this can be fixed for upcoming devnets.

**Comment by brusherru** [comment]
>Q: Is the happy flow of the basic use cases affected? e.g. Install smapp, run local node, start smashing?

It affects start smeshing within any custom directory (non-default ~/post) and, moreover, it prevents recreating valid PoST data even after a second Node start until the User deletes PoST data files manually.

E.G.
1. Start smeshing to the custom directory
2. PoST data created (but actually it is useless)
3. Rerun Node -> you'll get a new `key.bin` in the specified custom directory next to the previously created post data. However, this data will be "corrupted" (`ID` mismatch)
4. Setting up smeshing again in the same directory doesn't "heal" or replace these files. So every run will fail with the same thing, until...
5. Open this directory and delete post data files and run again.
6. Only now the smeshing data will be generated with proper ID.

This results in the following problems:
- Common users can't setup smeshing at all
- Advanced users can manage it, but probably they will smesh to the void until they restart the Node and find out the error. Since we're saying "run it 24/7" I can assume that most users will set up it once and then try to keep it working all the time. So probably they won't restart the Node at all.

Also, it blocks [setup post GPU](https://github.com/spacemeshos/smapp/pull/810) and as a consequence the release of Small 0.2. I can make a temporary workaround in Smapp, but I'm sure that this bug should be fixed within this repo ;)

@moshababo @avive @noamnelke please, take this issue into account 🙏

**Comment by avive** [comment]
@moshababo I hope that these requirements can be implemented soon https://github.com/spacemeshos/go-spacemesh/issues/2858#issuecomment-934164646

**Comment by noamnelke** [comment]
The problem description is mostly accurate. We generate a keypair to use as the `NodeID` upon startup, before it's required by the protocol. The protocol doesn't need the a keypair before generating PoST data. However, we've always used the generated keypair to identify the node in logs.

The advantage of using the same public key included in ATXs, blocks and Hare messages as a node identifier is that it's trivial to identify another node based messages they publish. If we use a separate identifier (that's kept private) to identify logs, another step is required when trying to access logs of a node that published some message: you need to locate a log message published by that node to map the public identity to the now-private `NodeID`.

Why do we even need an identity in logs? When we aggregate logs (in CI/testnet) we identify the pod/process that produced each log. The only place where this doesn't happen is in the app test, where multiple local nodes are spun up and they each just print their logs to `stdout`.

In the app test nodes are started with a config that points to the `PostDataDir` and a `StartSmeshing=true` value, so if nodes include the `PostID` in log messages once it's known that would mostly solve the app test issue. It would also mostly solve the issue of finding a node's logs in the aggregated logs scenario.

## What needs to change

1. No keypair should be generated unless/until `StartSmeshing=true`.
2. When the `PostDataDir` is changed, the `NodeID` should be loaded from that location if there's one there already. If there isn't - a new identity should be generated and stored there.

## Challenges

- We currently set the `NodeID` in the log object when creating it and before passing it into individual modules. We'll now need to enable globally setting this value somehow.
- Modules used for smeshing (block builder, hare, activation) today are initialized and started with the node, even if `StartSmeshing=false`. They are initialized with the `NodeID` and don't support changing the ID. We'll need to either not start those processes until `StartSmeshing=true` (and restart them when the key changes) or support changing the `NodeID` while they run. We don't have to choose the same solution for all modules, e.g. it makes sense to run Hare when the node starts up, because the same module processes Hare messages and generates them, but it makes more sense to start the activation module only when `StartSmeshing=true` because it also internally stores some identity-specific state, so supporting changing the key would be harder.

## Next steps

We should prioritize the long term fix and then I'll create detailed dev tasks. If we're going to fix this very soon then there's nothing else to do. If we can't afford to fix this soon, it may be worth implementing a temporary fix:

If the `PostDataDir` is changed at runtime, when a `NodeID` is already loaded, check if the new location has a `key.bin` in it. If it does and it's not the same key we have in memory - panic (since we don't support changing the loaded ID at runtime yet), if it doesn't - persist the currently loaded keypair in the new location, so it can be loaded at the next node startup.

cc: @avive @moshababo

**Comment by brusherru** [comment]
@noamnelke can't we cut the corners and do only this thing:

> 2. When the PostDataDir is changed, the NodeID should be loaded from that location if there's one there already. If there isn't - a new identity should be generated and stored there.

It seems to me that it will work (at least as a temporary solution for this issue and stop blocking release 0.2).

So right on calling `SmesherService.StartSmeshing` we have to check for the existence of `key.bin` in the specified directory.
- If it exists and differs from the NodeId which is already stored somewhere in the memory (so it was generated or loaded from the file, doesn't matter) — response with a proper error, store smeshing settings, but don't start creating PoST data immediately.
- Error like this: "Error: Found key.bin file in the PoST directory which differs from the current. Please, restart the Node to use this one or delete it if you want to create a new key."
- So after restarting Node it will: (a) use this key.bin, or (b) create a new one in the directory and start smeshing on startup.
- If it does not exist — create `key.bin` using `NodeId` from the memory, and start smeshing

This idea doesn't handle the error when post data was created, but `key.bin` was deleted. And there will be `ID mismatch` error. However, I think it is good and I don't sure that we can and we should protect from such cases. But what we really should do — ensure that `key.bin` is created at the right moment and used in PoS data creation as well.

--
Am I missing something important?

**Comment by noamnelke** [comment]
The minimal fix we can do with the largest impact, imo, is this:

When the SMApp calls `StartSmeshing`, copy `key.bin` from the previous to the new location. If a file with that name exists there already - panic. This alone will fix the problem with the way the SMApp uses the node.

What won't work? If a user needs to set up SMApp a second time when their PoST data is already initialized, the node will fail. This is a less common case, so I can live with immediately fixing the common case and fixing this one later on.

@brusherru @moshababo wdyt?

**Comment by avive** [comment]
> What won't work? If a user needs to set up SMApp a second time when their PoST data is already initialized, the node will fail. This is a less common case, so I can live with immediately fixing the common case and fixing this one later on.

This is actually quite common - in the case of installing smapp on a new devnet/testnet and specifying a folder for post that has old data. It happens to me all the time and currently Smapp displays the id mismatch stuff.

To clarify, at minimum, for the short-term we want to make sure that at least this basic use case works:
1. Smapp session - setup post and start smeshing
2. Restart smapp - smeshing should start automatically (smapp modified the config file) and with same 'nodeId' as prev session.

It is my understanding that this scenario doesn't work now.

For the longer term, we want to implement these requirements. If they are not acceptable to everyone then let's please discuss them: https://github.com/spacemeshos/go-spacemesh/issues/2858#issuecomment-934164646

**Comment by noamnelke** [comment]
I see, however my point still stands. The node doesn't currently support changing the `NodeID` at runtime. We pass this value to many different modules at startup and the node is not currently designed to support updating it. I agree that this design should be changed, but this is not a quick fix. Doing what I suggested (moving `key.bin` if the new folder is empty) is easy to do and will solve a more common use-case than yours.

I still suggest doing that now and doing the longer fix later. What you're saying just means "later" should probably be soon.

**Comment by avive** [comment]
What about the more basic use case that doesn't work outlined in my comment?

**Comment by noamnelke** [comment]
This is what my suggestion fixes

**Comment by avive** [comment]
But Smapp doesn't call 'Start Smeshing' on 2nd app session - it just gives the node a modified config file with the same params from the first session to start with.... The current pattern is:

1. App session start and post was not previously setup by user. Node is started w/o any post data and startSmeshing is set to false.
2. User sets up post: Smapp uses the API to init post and to start smeshing. Smapp modified the node's config file with the user-set data so node will start on future sessions using this config.
3. New app session - node is started with the modified config file and so it is smeshing on startup and no api calls are used.

**Comment by avive** [comment]
Also, regarding: 

> What won't work? If a user needs to set up SMApp a second time when their PoST data is already initialized, the node will fail. This is a less common case, so I can live with immediately fixing the common case and fixing this one later on.

Can you plz clarify this scenario a bit? What do you mean by setting up Smapp the 2nd time? Do you mean installing Smapp and specifying a post data dir that was previously used (in an older smapp session, possibly on an older net)?

**Comment by brusherru** [comment]
> If a user needs to set up SMApp a second time when their PoST data is already initialized, the node will fail

It also will cause Node panic on "Modify PoST data" (E.G. changing size).

So in this case I can suggest making a kludge in Smapp, until it will be fixed well in go-sm:
- Before starting smeshing Smapp will check for the existence of the `key.bin` file in the specified directory:
    - If it is not there — just copy it from the default path (`~/post/data/key.bin`)
    - If it is there — check that this is the same file using shasum or something similar. If it differs — prompt User what to do with it (overwrite / use it). In case of selecting "use it" — it should update the config and restart the Node without calling `StartSmeshing` (so PoST data would not be created until the Node restarts and starts smeshing by the config, not GRPC call).

**Comment by noamnelke** [comment]
@avive Let me explain:

1. User starts SMapp for the first time. The SMApp passes a temporary `PostDatadir` to the node and the node creates `key.bin` there.
2. During runtime, the user "sets up smeshing", causing the SMApp to call `StartSmeshing` with a new `PostDatadir`. The node already has a key loaded (the one stored in temporary `PostDatadir`) and it uses that key to initialize PoST.
3. The user then restarts the SMApp (and node) and the new, correct `PostDatadir` is in the config. The node looks for `key.bin` there, but it doesn't exist, so a new one is created, but doesn't match the PoST data.

What I'm suggesting is that in step 2 `key.bin` will be transferred over, so that in step 3 it can be loaded from the correct location.

**Comment by noamnelke** [comment]
@brusherru I suggested to do check first if the datadir changed...

The "kludge" you suggested is perfect. Then we don't need a short term fix in the node.

**Comment by noamnelke** [comment]
You can possibly do even better - when the user sets up smeshing, if the new folder is empty, copy the file right there and then.

**Comment by noamnelke** [comment]
And there's also no need to ask the user - if they set up smeshing with a folder that has a different key in it - restart the node. I can't see a case where the user will want to do something else.

**Comment by avive** [comment]
All-right, so it sounds like we have a plan for a short-term kludge to fix the basic scenario.

**Comment by noamnelke** [comment]
it fixes both scenarios, but it's less efficient, since you must restart the node when setting up SMApp with an existing PoST dir. This means that sync+gossip will restart.

In the long term I still think we need the fix I suggested, but Kirill's suggestion makes it much less urgent.

**Comment by brusherru** [comment]
So I still can check for shashum equality to avoid excessive restarts (E.G. on Modify PoST data).
Just trying to imagine will it introduce other problems (syncing is always a problem) 🤔

**Comment by noamnelke** [comment]
I think the solution is good.

Having a `key.bin` which matches the PoST data in the `PostDatadir` is the desired, consistent behavior. I can't think of any issue this can cause. Restarting the node so it starts up with the correct `PostDatadir` (and the correct `key.bin` in it) is consistent and should work.

**Comment by noamnelke** [comment]
When the node supports changing the `NodeID` at runtime, we can make this flow more efficient (specifically, no need to restart), but it won't functionally change anything, compared to the kludge.

**Comment by brusherru** [comment]
Cool. Let's summarize:
1. Make a temporary kludge in Smapp (urgent: https://github.com/spacemeshos/smapp/issues/823 ):
    - Before starting smeshing Smapp will check for the existence of the `key.bin` file in the specified directory:
       - If it is not there — just copy it from the default path (`~/post/data/key.bin`)
       - If it is there — check that this is the same file as in the default path (`~/post/data/key.bin`) using shasum. If it differs — update the config and restart the node. If it is the same — just call `StartSmeshing`.
       - If `~/post/data/key.bin` not found — update the config and restart the node as well
2. Make a robust fix in go-spacemesh to accept requirements: https://github.com/spacemeshos/go-spacemesh/issues/2858#issuecomment-934164646 

If it sounds good, I'll create an issue for Smapp and then paste a link here.
About the issue related to go-sm, I propose to summarize everything in the new issue, post the link here and close this one :)

**Comment by noamnelke** [comment]
💯 

I'll add creating an issue for `go-spacemesh` to my todo list.

**Comment by brusherru** [comment]
Created issue for Smapp: https://github.com/spacemeshos/smapp/issues/823

**Comment by brusherru** [comment]
Folks, we have had this issue for a long time.
We already have a working solution with "kludge" in Smapp (https://github.com/spacemeshos/go-spacemesh/issues/2858#issuecomment-960733520). The main question is — do we want to do a second part (on Node side) and when. It does not seem urgent/critical/priority, so probably we should just close the issue (or add an "on-hold" label).

@noamnelke @avive
